File: clojure/anagram/src/anagram.clj
(ns anagram 
  (:require [clojure.string :as str]))
  
(defn anagram?
  [word tested-word] 
  (let [word (str/lower-case word) 
        tested-word (str/lower-case tested-word)] 
        (and (not= word tested-word) 
          (= (frequencies word) 
             (frequencies tested-word)))))  

(defn anagrams-for
  [word candidates] 
  (filter #(anagram? word %) candidates)) 

File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.List
import Data.Char

anagramsFor :: String -> [String] -> [String]
anagramsFor word = filter (predicate word)

predicate :: String -> String -> Bool
predicate word possibleAnagram =
  let
    word' = map toLower word
    possibleAnagram' = map toLower possibleAnagram
  in
    word' /= possibleAnagram' && sort word' == sort possibleAnagram'
